### DISCUSSION TITLE
Store unique device ID, write directly to flash memory (using SWD)

### DISCUSSION THREAD

**Discussion by RobertMorton56** [earlephilhower/arduino-pico#1391] [Q&A] [answered]
Hi,

I can't seem to find a solution from reading the docs and the issues section so I'll ask here...

Is there an easy reliable way to write a unique uint32_t to the RP2040 flash memory, from the command line (using OpenOCD or similar), then have the RP2040 read this uint32_t to use it as a UID during operation. I'd like to have a flexible way of overwriting UIDs on devices from the command line. I imagine I could use the EEPROM library to read that section of flash during normal operation but there's no documentation that gives me a hint of how do to what I want to do. Exact memory locations etc...

Thanks in advance, and thanks for a great core.
Best,
Bob

**Comment by earlephilhower** [comment]
You need to run an application on the RP2040 to actually erase and write to the flash.  OpenOCD just provides a way for GDB to get into the core.

Each Pico already has a unique multibyte ID, see https://www.raspberrypi.com/documentation/pico-sdk/high_level.html#pico_unique_id

**Comment by RobertMorton56** [comment]
Apologies for not putting this question in discussions and thanks for your answer. I understand the ID is available to some flash chips but I need things to be a little more flexible. 

To be a little more specific: I'm using the W25Q128JV flash chip on a custom RP2040 board. So could I flash firmware, then at some point after the initial flashing write some uint32_t values to the flash at some specific location. During code execution then read that location (perhaps with EEPROM lib?) for some application specific read only variables, GUIDs etc. 

Really I'm just trying to store some application specific read-only data for use in the firmware. ESP32 has NVS, STM32 has access to the linker script. What would you recommend for this functionality on the RP2040 core?

**Reply by maxgerhardt** [reply]
The EEPROM.h implementation here is also just accessing the last flash sector (4096 bytes) in a pretty direct way, see [source](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/EEPROM/src/EEPROM.cpp). Meaning if you direct OpenOCD to do a flash write to that flash address, you will end up being able to read it with the library.

**Reply by RobertMorton56** [reply]
Great, that's what I was hoping for. Thanks for that. So this is a fairly robust way of doing this? No potential pitfalls?

**Reply by earlephilhower** [reply]
Just be aware it's not a real EEPROM.  So if you need to update something in-situ it's read-modify-erase-write (taken care of by the `EEPROM` library).

If you're updating things frequently, I would suggest using a `File` in `LittleFS` since that will wear-level writes across the entire space, not constantly re-write a single flash sector.

**Reply by RobertMorton56** [reply]
Understood. Thankfully I'll only be writing a handful of times - and outside normal operation of the device. Thanks again for the help, so far this core has been wonderful to work with.

**Reply by RobertMorton56** [reply]
> The EEPROM.h implementation here is also just accessing the last flash sector (4096 bytes) in a pretty direct way, see [source](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/EEPROM/src/EEPROM.cpp). Meaning if you direct OpenOCD to do a flash write to that flash address, you will end up being able to read it with the library.

Perhaps you can correct me here. I've created a new board via the "Adding a New Board" docs. And because I'm using a 16MB flash (W25Q128JVSIQ) it spits out:

`newboard.menu.flash.16777216_0.build.eeprom_start=285208576`

That's address 0x10FFF000

So I attempt to flash using my picoprobe like so:

openocd -f interface/cmsis-dap.cfg -f target/rp2040.cfg -c "adapter speed 5000" -c "program testData.bin reset exit 0x10FFF000"

And it's telling me:

Warn : no flash bank found for address 0x10fff000

I know I'm doing something stupid but I can't see it. If you had a concrete example of how to use openocd to write either a binary or even just a few bytes to the start of EEPROM memory I'd be very grateful. Thanks.

**Reply by earlephilhower** [reply]
Off the to of my head, drop the `0x10` prefix and try uploading to `0xfff000` instead.  The memory mapped space for flash is not always the same address used for QSPI erase/writes in many micros.  You could also look in the `EEPROM.cpp` file to see what gymnastics it goes though to get the erase/program address to call the ROM APIs.

**Reply by RobertMorton56** [reply]
> Off the to of my head, drop the `0x10` prefix and try uploading to `0xfff000` instead. The memory mapped space for flash is not always the same address used for QSPI erase/writes in many micros. You could also look in the `EEPROM.cpp` file to see what gymnastics it goes though to get the erase/program address to call the ROM APIs.

Thanks, `0xfff000` and even `0xffe000` still throw `Warn : no flash bank found for address 0x00fff000`. I'll dig into the EEPROM code and see if I can figure it out.

**Reply by maxgerhardt** [reply]
Try adding a `init; ` before `program` to make sure it's enumerated the flash banks properly. Or even `flash probe;`

**Reply by RobertMorton56** [reply]
The openocd target config file for the rp2040 (target/rp2040.cfg) seems to hard-code the flash memory size to 2MB?

`#set _FLASHSIZE 0x200000
set _FLASHSIZE 0xFFF000`

As you can see I changed it to 16MB and now openocd doesn't complain about the missing banks. Still, I can't seem to change the values of EEPROM from outside of the core. I'll continue digging.

**Reply by maxgerhardt** [reply]
Oh if that's true then I'll definitely need to adapt the update logic to pass a `-c set _FLASHSIZE <flash size>` argument in the OpenOCD invocation when uploading to boards with bigger flash. Good catch.

**Reply by RobertMorton56** [reply]
Seems like it, I just checked my local files, but [the source](https://github.com/raspberrypi/openocd/blob/rp2040/tcl/target/rp2040.cfg#L37) shows the same.

**Reply by RobertMorton56** [reply, marked as answer]
Think I have it, the openocd config specifies the flash size as `set _FLASHSIZE 0x200000` but of course the flash size has to be the entire flash device memory `set _FLASHSIZE 0x1000000`. After updating this I can run `openocd -f interface/cmsis-dap.cfg -f target/rp2040.cfg -c "adapter speed 5000" -c "program testData.bin reset exit 0x10FFF000"` and voilà ... I can now read that testData.bin data in the EEPROM library of rp2040 (EEPROM.read()).
